src/parser/RegExpParser.mts

```diff
@@ -321,70 +401,152 @@ export class RegExpParser {
+            production: 'Group',
+            leftCapturingParenthesesBefore: this.leftCapturingParenthesesBefore - 1,
+            GroupSpecifier: name,
+            Disjunction: this.parseDisjunction(),
+          };
+          this.getAllGroupsWithName(name).push(node);
+          this.capturingGroups.push({ GroupName: name, position: groupNamePos });
         } else {
-          node.GroupSpecifier = this.parseGroupName();
-        }
-      }
-      if (node.capturing) {
-        this.capturingGroups.push(node);
-      }
-      if (node.GroupSpecifier) {
-        if (this.groupSpecifiers.has(node.GroupSpecifier)) {
-          this.raise(`Duplicate group specifier '${node.GroupSpecifier}'`);
+          const { PlusModifiers, MinusModifiers } = this.parseAtomModifiers();
+          node = {
+            type: 'Atom',
+            production: 'Modifier',
+            leftCapturingParenthesesBefore: this.leftCapturingParenthesesBefore,
+            AddModifiers: PlusModifiers,
+            RemoveModifiers: MinusModifiers,
+            Disjunction: this.parseDisjunction(),
+          };
         }
-        this.groupSpecifiers.set(node.GroupSpecifier, node.capturingParenthesesBefore);
+      } else {
+        this.leftCapturingParenthesesBefore += 1;
+        node = {
+          type: 'Atom',
+          production: 'Group',
+          leftCapturingParenthesesBefore: this.leftCapturingParenthesesBefore - 1,
+          GroupSpecifier: undefined,
+          Disjunction: this.parseDisjunction(),
+        };
+        this.capturingGroups.push({ GroupName: undefined, position: this.position });
       }
-      node.Disjunction = this.parseDisjunction();
       this.expect(')');
-      node.enclosedCapturingParentheses = this.capturingGroups.length - node.capturingParenthesesBefore - 1;
       return node;
     }
     if (this.test('[')) {
       return {
         type: 'Atom',
+        production: 'CharacterClass',
         CharacterClass: this.parseCharacterClass(),
       };
     }
     if (isSyntaxCharacter(this.peek())) {
-      this.raise(`Expected a PatternCharacter but got ${this.peek()}`);
+      this.raise(`Expected a character but got ${this.peek()}`);
     }
     return {
       type: 'Atom',
+      production: 'PatternCharacter',
       PatternCharacter: this.parseSourceCharacter(),
     };
   }
 
+  // WhatWeAreParsingHere :: (used in Atom, `<` is for named capture groups)
+  //   [empty] [lookahead = `:` or `<`]
+  //   RegularExpressionModifiers [lookahead = `:` or `<`]
+  //   RegularExpressionModifiers `-` RegularExpressionModifiers [lookahead = `:` or `<`]
+  //
+  // RegularExpressionModifiers ::
+  //   [empty]
+  //   RegularExpressionModifiers RegularExpressionModifier
+  //
+  // RegularExpressionModifier :: one of `i` `m` `s`
+  private parseAtomModifiers(): {
+    PlusModifiers: ParseNode.RegExp.RegularExpressionModifier[] | undefined;
+    MinusModifiers: ParseNode.RegExp.RegularExpressionModifier[] | undefined;
+    } {
```

Comment:
I'm not a big fan of inline object literal types as return types.

Suggestion:
```typescript
private parseAtomModifiers(): Record<"PlusModifiers" | "MinusModifiers", ParseNode.RegExp.RegularExpressionModifier[] | undefined> {
```